File: WalletApi/WalletApi/Common/ApiEnvelope.cs
namespace WalletApi.Common;

/// <summary>
/// Standardized API response wrapper. Every response — success or failure — uses this envelope.
/// </summary>
public sealed class ApiEnvelope<T>
{
    public bool Success { get; private init; }
    public T? Data { get; private init; }
    public IReadOnlyList<string>? Errors { get; private init; }
    public string? TraceId { get; private init; }
    public DateTime Timestamp { get; private init; }

    public static ApiEnvelope<T> Ok(T data, string? traceId = null) => new()
    {
        Success = true,
        Data = data,
        TraceId = traceId,
        Timestamp = DateTime.UtcNow
    };

    public static ApiEnvelope<T> Fail(IReadOnlyList<string> errors, string? traceId = null) => new()
    {
        Success = false,
        Errors = errors,
        TraceId = traceId,
        Timestamp = DateTime.UtcNow
    };

    public static ApiEnvelope<T> Fail(string error, string? traceId = null)
        => Fail([error], traceId);
}

File: wallet_app/lib/core/network/api_envelope.dart
library;

class ApiEnvelope<T> {
  final bool success;
  final T? data;
  final List<String>? errors;
  final String? traceId;
  final DateTime? timestamp;

  const ApiEnvelope({
    required this.success,
    this.data,
    this.errors,
    this.traceId,
    this.timestamp,
  });

  factory ApiEnvelope.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic json)? dataParser,
  ) {
    return ApiEnvelope<T>(
      success: json['success'] as bool? ?? false,
      data: json['data'] != null && dataParser != null
          ? dataParser(json['data'])
          : null,
      errors: (json['errors'] as List<dynamic>?)
          ?.map((e) => e.toString())
          .toList(),
      traceId: json['traceId'] as String?,
      timestamp: json['timestamp'] != null
          ? DateTime.tryParse(json['timestamp'] as String)
          : null,
    );
  }
}
